### PR TITLE
Private/shubham/grow 144

### PR DESCRIPTION
**Pull request by Shubhamag12**
## ISSUE(S):
[Link to jira](https://platform9.atlassian.net/browse/GROW-144)

## SUMMARY
Added support to pass environment variables through a file.

## ISSUE TYPE
- 🎁 New feature (non-breaking change which adds functionality)

## IMPACTED FEATURES/COMPONENTS:
appctl deploy command

#### Manual

- When file_flag (-f) is passed.
- When invalid file_path is given.
- When same environment variable is passed in both (in env file and in command line)
- When different environment variables is passed in flags, we will take union of both
- With -e flag
- With k-native image file 

##### Testing Output (Snapshot/file)
file_flag

invalid_file-path

same_env_both

union

with_e-flag

with_knative

### CODE REVIEW

File: pkg/appAPIs/appAPIs.go [hunk 1]
```diff
@@ -129,15 +131,50 @@ func (cli_api *AppAPI) createAppAPI(createInfo string, token string) ([]byte, er
...
+		for key, _ := range envMap {
+			_, found := sliceMap[key]
+			if found {
+				return fmt.Errorf("Duplicate environment variable: %v found. Either remove it from env file or from command line.", key)
+			}
+		}
+
+		slice := append(sliceFromEnv, ",")
+		slice = append(slice, sliceFromEnvFile...)
+		fmt.Printf("%s", slice)
```
Comment (AnirudhPokala) [inline pkg/appAPIs/appAPIs.go]: Why do we need to print env variables?

**Review by AnirudhPokala — CHANGES_REQUESTED** [review]

File: cmd/appDeploy.go [hunk 2]
```diff
@@ -79,6 +80,7 @@ func init() {
 	appCmdDeploy.Flags().StringVarP(&deployApp.userName, "username", "u", "", "Username of private container registry")
 	appCmdDeploy.Flags().StringVarP(&deployApp.password, "password", "P", "", "Password of private container registry")
 	appCmdDeploy.Flags().StringArrayVarP(&deployApp.env, "env", "e", nil, "Environment variable to set, as key=value pair")
+	appCmdDeploy.Flags().StringVarP(&deployApp.envFilePath, "envPath", "f", "", "File path for Enviroment variables, as line separated key=value pair")
```
Comment (AnirudhPokala) [inline cmd/appDeploy.go]: NiT: "Path to the environment variables file."
Also lets include format of values something like
"Values in the .env file should be formatted as a line separated Key=Value pair"

File: pkg/appAPIs/appAPIs.go (hunk 1 above)
Comment (Shubhamag12) [reply pkg/appAPIs/appAPIs.go]: I added for my reference, forgot to remove.
Will apply this change

**Comment by AnirudhPokala** [comment]
Let's also add an example of how to use `--envPath` flag in deploy_examples.

File: pkg/appAPIs/appAPIs.go [hunk 3]
```diff
@@ -129,15 +131,50 @@ func (cli_api *AppAPI) createAppAPI(createInfo string, token string) ([]byte, er
...
+	// if both are provided then we will take union of two
+	if env != nil && envFilePath != "" {
+		sliceFromEnv, sliceMap := genEnvSlice(env)
+		sliceFromEnvFile, envMap, err := GetSliceFromEnvFile(envFilePath)
+
+		if err != nil {
+			return fmt.Errorf("%s", err)
+		}
+
+		// if a same key is found in both (envFile and env from command line, we will throw an error)
```
Comment (AnirudhPokala) [inline pkg/appAPIs/appAPIs.go]: Let's add a detailed comments before all  cases. 
NiT: If the same key is found in both the `.env` file passed through --envPath and the --env flag.

File: pkg/appAPIs/appAPIs.go [hunk 4]
```diff
@@ -407,19 +444,46 @@ func Login(token string) error {
...
+	if envFilePath != "" {
+		envFile, err := os.Open(envFilePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Error opening the env file. Please make sure that file path: %s is valid.", envFilePath)
+		}
+		defer envFile.Close()
+		scanner := bufio.NewScanner(envFile)
+		for scanner.Scan() {
+			text := scanner.Text()
+			splitEnv := strings.Split(text, "=")
```
Comment (AnirudhPokala) [inline pkg/appAPIs/appAPIs.go:490]: What if data in `.env` file is not of specified format? i.e not of Key=Value. Shouldn't we error out with a proper error message.

Comment (Shubhamag12) [reply pkg/appAPIs/appAPIs.go:490]: handled this one
wrong_format

File: pkg/appAPIs/appAPIs.go [hunk 5]
```diff
@@ -407,19 +455,51 @@ func Login(token string) error {
...
+	var envSlice []string
+	envMap := make(map[string]string)
+	if envFilePath != "" {
+		envFile, err := os.Open(envFilePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Error opening the env file. Please make sure that file path: %s is valid.", envFilePath)
+		}
+		defer envFile.Close()
+		scanner := bufio.NewScanner(envFile)
+		regex, _ := regexp.Compile("[[:alnum:]]+=[[:alnum:]]+")
```
Comment (AnirudhPokala) [inline pkg/appAPIs/appAPIs.go]: Let's add this Regex value into constants file.

File: cmd/appDeploy.go [hunk 6]
```diff
@@ -45,6 +45,14 @@ var deployExample = `
   # Deploy an app using app-name, container image and pass environment variables and set port where application listens on.
   appctl deploy -n <appname> -i <image> -e key1=value1 -e key2=value2 -p <port>
   Ex: appctl deploy -n hello -i gcr.io/knative-samples/helloworld-go -e TARGET="appctler" -p 7893
+
+  # Deploy an app using app-name, container image and pass environment variables through a file and set port where application listens on.
+  appctl deploy -n <appname> -i <image> -f <env-file-path> -p <port>
+  Ex: appctl deploy -n hello -i gcr.io/knative-samples/helloworld-go -f /Users/user/variables.env -p 7893
+
+  # Deploy an app using app-name, container image and pass environment variables through a file or pass through command line and set port where application listens on.
```
Comment (AnirudhPokala) [inline cmd/appDeploy.go]: I think its ... pass environment variables through a file `and` pass through command line instead of `or`

**Review by AnirudhPokala — DISMISSED** [review]
LGTM

File: pkg/appAPIs/appAPIs.go [hunk 7]
```diff
@@ -129,15 +131,60 @@ func (cli_api *AppAPI) createAppAPI(createInfo string, token string) ([]byte, er
...
+		} else {
+			createInfo = fmt.Sprintf(`{"name":"%s", "image":"%s", "username":"%s", "password":"%s", "envs": %v}`, name, image, username, password, envSlice)
+		}
+	} else if envFilePath != "" {
+		envSlice, _, err := GetSliceFromEnvFile(envFilePath)
+		if err != nil {
+			return fmt.Errorf("%s", err)
+		} else if port != "" {
+			createInfo = fmt.Sprintf(`{"name":"%s", "image":"%s", "username":"%s", "password":"%s", "port": "%s", "envs": %v}`, name, image, username, password, port, envSlice)
 		} else {
```
Comment (anupbarve) [inline pkg/appAPIs/appAPIs.go]: This piece of code has become effectively unreadable.
Could you place it in smaller functions for the reader of the code to make sense out of it ?

File: cmd/appDeploy.go [hunk 8]
```diff
@@ -45,6 +45,14 @@ var deployExample = `
...
   appctl deploy -n <appname> -i <image> -e key1=value1 -e key2=value2 -p <port>
   Ex: appctl deploy -n hello -i gcr.io/knative-samples/helloworld-go -e TARGET="appctler" -p 7893
+
+  # Deploy an app using app-name, container image and pass environment variables through a file and set port where application listens on.
+  appctl deploy -n <appname> -i <image> -f <env-file-path> -p <port>
+  Ex: appctl deploy -n hello -i gcr.io/knative-samples/helloworld-go -f /Users/user/variables.env -p 7893
+
+  # Deploy an app using app-name, container image and pass environment variables through a file and pass through command line and set port where application listens on.
+  appctl deploy -n <appname> -i <image> -f <env-file-path> -e key1=value1 -e key2=value2 -p <port>
+  Ex: appctl deploy -n hello -i gcr.io/knative-samples/helloworld-go -f /Users/user/variables.env -e TARGET="appctler" -p 7893
```
Comment (anupbarve) [inline cmd/appDeploy.go:55]: Need to make appropriate changes in README.md

**Review by anupbarve — CHANGES_REQUESTED** [review]

File: README.md [hunk 9]
```diff
@@ -219,6 +228,13 @@ Appctl supports multiple ```--env``` variables
 ./appctl deploy --app-name <name> --image <docker-image path> --env key1=value1 --env key2=value2
 ```
 
+Appctl supports variables passed through a env file
+
+```sh
+./appctl deploy --app-name <name> --image <docker-image path> -f <env-file path>
+
+Example:
+./appctl deploy -n hello -i gcr.io/knative-samples/helloworld-go -f /Users/user/variables.env
```
Comment (anupbarve) [inline README.md:237]: - Add an example of how variables.env would look like.
- Add the behaviour when env variables are passed both via file and command line.

**Review by anupbarve — APPROVED** [review]

**Review by AnirudhPokala — APPROVED** [review]
LGTM

**Pull request merged**